tensorflow_addons/optimizers/moving_average.py

```diff
@@ -69,109 +71,30 @@ def __init__(self,
-                ma_op = self._ema.apply(var_list)
-        else:
-            ma_op = self._ema.apply(var_list)
+        self._set_hyper("average_decay", average_decay)
 
-        return tf.group(train_op, ma_op, name="train_with_avg")
+    def average_op(self, var, average_var):
+        decay = self._get_hyper('average_decay', tf.dtypes.float32)
+        return assign_moving_average(average_var, var, decay, False)
 
     def get_config(self):
         config = {
-            'optimizer': tf.keras.optimizers.serialize(self._optimizer),
             'average_decay': self._serialize_hyperparameter('average_decay'),
             'num_updates': self._num_updates,
-            'sequential_update': self._sequential_update
         }
         base_config = super(MovingAverage, self).get_config()
         return dict(list(base_config.items()) + list(config.items()))
 
-    @classmethod
-    def from_config(cls, config, custom_objects=None):
-        optimizer = tf.keras.optimizers.deserialize(
-            config.pop('optimizer'), custom_objects=custom_objects)
-        return cls(optimizer, **config)
-
-    def assign_average_vars(self, var_list):
-        """Assign variables in var_list with their respective moving averages.
-
-        Example:
-        ```python
-        model = tf.Sequential([...])
-        opt = tfa.optimizers.MovingAverage(
-            tf.keras.optimizers.SGD(lr=2.0), 0.5)
-
-        model.compile(opt, ...)
-        model.fit(x, y, ...)
-
-        # Update the weights to their mean before saving
-        opt.assign_average_vars(model.variables)
-
-        model.save('model.h5')
-        ```
-        """
-        assign = tf.group([v.assign(self._ema.average(v)) for v in var_list])
-        return assign
-
-    @property
-    def weights(self):
-        return self._optimizer.weights
-
-    def _resource_apply_dense(self, grad, var):
-        return self._optimizer._resource_apply_dense(grad, var)  # pylint: disable=protected-access
-
-    def _resource_apply_sparse_duplicate_indices(self, grad, var, indices):
-        return self._optimizer._resource_apply_sparse_duplicate_indices(  # pylint: disable=protected-access
-            grad, var, indices)
-
-    def _resource_apply_sparse(self, grad, var, indices):
-        return self._optimizer._resource_apply_sparse(grad, var, indices)  # pylint: disable=protected-access
-
-    @property
-    def learning_rate(self):
-        return self._optimizer._get_hyper('learning_rate')
-
-    @learning_rate.setter
-    def learning_rate(self, learning_rate):
-        self._optimizer._set_hyper('learning_rate', learning_rate)
-
-    @property
-    def lr(self):
-        return self.learning_rate
-
-    @lr.setter
-    def lr(self, lr):
-        self.learning_rate = lr
+    def _create_slots(self, var_list):
+        self._optimizer._create_slots(var_list=var_list)  # pylint: disable=protected-access
+        for var in var_list:
+            self.add_slot(var, 'average', var.read_value())
```

Comment:
Is using `var.read_value()` the right way to initialize a slot value?

Reply:
Because this requires us to run a variables initializer twice, once when the model is defined and once after `model.compile()`

Reply:
I wasn't able to find a better alternative. `.value()` may cause issues for distributed strategies, and didn't see an initializer from existing variable elsewhere.